tokio-util/src/sync/cancellation_token.rs

```diff
@@ -99,6 +99,18 @@ pin_project! {
     }
 }
 
+pin_project! {
+    /// A Future that is resolved once the corresponding [`CancellationToken`]
+    /// is cancelled or a given Future gets resolved. It is biased towards the
+    /// Future completion.
+    #[must_use = "futures do nothing unless polled"]
+    struct RunUntilCancelledFuture<'a, F: Future> {
+        cancellation_token: &'a CancellationToken,
+        #[pin]
+        future: F,
+    }
+}
```

Comment:
One option is to declare the struct _inside_ of `run_until_cancelled`. That way, it's clear that it's only here for that method.

Reply:
Although I find it strange to have the `impl`-block of the future type defined inside of the function as well, I'm fine with that.